ElasticsearchExporterFaultToleranceIT.shouldExportEvenIfElasticNotInitiallyReachable flaky

**Summary**

- How often does the test fail? once so far
- Does it block your work? no
- Do we suspect that it is a real failure? unknown

**Failures**
(Currently this information is not available)

Happened again in https://ci.zeebe.camunda.cloud/job/camunda-cloud/job/zeebe/job/staging/490/

```
[2021-06-01T12:57:37.487Z] [WARNING] io.camunda.zeebe.exporter.ElasticsearchExporterFaultToleranceIT.shouldExportEvenIfElasticNotInitiallyReachable
[2021-06-01T12:57:37.487Z] [ERROR]   Run 1: ElasticsearchExporterFaultToleranceIT.shouldExportEvenIfElasticNotInitiallyReachable:37->lambda$shouldExportEvenIfElasticNotInitiallyReachable$1:37 
[2021-06-01T12:57:37.487Z] Expecting value to be true but was false expected:<[tru]e> but was:<[fals]e>
[2021-06-01T12:57:37.487Z] [INFO]   Run 2: PASS
[2021-06-01T12:57:37.487Z] [INFO] 
[2021-06-01T12:57:37.487Z] [INFO] 
[2021-06-01T12:57:37.487Z] [WARNING] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Flakes: 1
```

```
[2021-06-01T12:56:57.244Z] [ERROR] io.camunda.zeebe.exporter.ElasticsearchExporterFaultToleranceIT.shouldExportEvenIfElasticNotInitiallyReachable  Time elapsed: 49.036 s  <<< FAILURE!
[2021-06-01T12:56:57.244Z] org.junit.ComparisonFailure: 
[2021-06-01T12:56:57.244Z] 
[2021-06-01T12:56:57.244Z] Expecting value to be true but was false expected:<[tru]e> but was:<[fals]e>
[2021-06-01T12:56:57.244Z] 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[2021-06-01T12:56:57.244Z] 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[2021-06-01T12:56:57.244Z] 	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[2021-06-01T12:56:57.244Z] 	at io.camunda.zeebe.test.util.TestUtil$Invocation.until(TestUtil.java:91)
[2021-06-01T12:56:57.244Z] 	at io.camunda.zeebe.test.util.TestUtil$Invocation.until(TestUtil.java:64)
[2021-06-01T12:56:57.244Z] 	at io.camunda.zeebe.test.util.TestUtil.waitUntil(TestUtil.java:36)
[2021-06-01T12:56:57.244Z] 	at zeebe.elasticsearch.exporter@1.1.0-SNAPSHOT/io.camunda.zeebe.exporter.ElasticsearchExporterFaultToleranceIT.lambda$shouldExportEvenIfElasticNotInitiallyReachable$1(ElasticsearchExporterFaultToleranceIT.java:37)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
[2021-06-01T12:56:57.244Z] 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
[2021-06-01T12:56:57.244Z] 	at io.camunda.zeebe.test.util.stream.StreamWrapper.forEach(StreamWrapper.java:175)
[2021-06-01T12:56:57.244Z] 	at zeebe.elasticsearch.exporter@1.1.0-SNAPSHOT/io.camunda.zeebe.exporter.ElasticsearchExporterFaultToleranceIT.shouldExportEvenIfElasticNotInitiallyReachable(ElasticsearchExporterFaultToleranceIT.java:37)
[2021-06-01T12:56:57.244Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2021-06-01T12:56:57.244Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[2021-06-01T12:56:57.244Z] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2021-06-01T12:56:57.244Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[2021-06-01T12:56:57.244Z] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
[2021-06-01T12:56:57.244Z] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[2021-06-01T12:56:57.244Z] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
[2021-06-01T12:56:57.244Z] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[2021-06-01T12:56:57.244Z] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
[2021-06-01T12:56:57.244Z] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
[2021-06-01T12:56:57.244Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
[2021-06-01T12:56:57.244Z] 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
[2021-06-01T12:56:57.244Z] 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
[2021-06-01T12:56:57.244Z] 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
[2021-06-01T12:56:57.245Z] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
[2021-06-01T12:56:57.245Z] 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:364)
[2021-06-01T12:56:57.245Z] 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
[2021-06-01T12:56:57.245Z] 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:237)
[2021-06-01T12:56:57.245Z] 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:158)
[2021-06-01T12:56:57.245Z] 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
[2021-06-01T12:56:57.245Z] 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
[2021-06-01T12:56:57.245Z] 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
[2021-06-01T12:56:57.245Z] 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
```

--- a/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/AbstractElasticsearchExporterIntegrationTestCase.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/AbstractElasticsearchExporterIntegrationTestCase.java
@@ -19,33 +19,36 @@ import io.camunda.zeebe.exporter.dto.GetSettingsForIndicesResponse.IndexSettings
 import io.camunda.zeebe.exporter.util.ElasticsearchContainer;
 import io.camunda.zeebe.exporter.util.ElasticsearchNode;
 import io.camunda.zeebe.exporter.util.it.ExporterIntegrationRule;
+import io.camunda.zeebe.exporter.util.it.NonStartableBrokerRule;
 import io.camunda.zeebe.protocol.record.Record;
 import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.ValueTypeMapping;
 import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
 import java.io.IOException;
 import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 import org.awaitility.Awaitility;
 import org.elasticsearch.client.Request;
+import org.elasticsearch.client.ResponseException;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Rule;
-import org.junit.rules.TemporaryFolder;
 
 public abstract class AbstractElasticsearchExporterIntegrationTestCase {
   private static final ObjectMapper MAPPER = new ObjectMapper();
-
-  @Rule public final TemporaryFolder temporaryFolder = new TemporaryFolder();
 
   @Rule
   public final RecordingExporterTestWatcher recordingExporterTestWatcher =
       new RecordingExporterTestWatcher();
 
-  protected final ExporterIntegrationRule exporterBrokerRule = new ExporterIntegrationRule();
-
   protected ElasticsearchNode<ElasticsearchContainer> elastic;
   protected ElasticsearchExporterConfiguration configuration;
   protected ElasticsearchExporterFaultToleranceIT.ElasticsearchTestClient esClient;
+
+  private final ExporterIntegrationRule exporterIntegrationRule = new ExporterIntegrationRule();
+  protected final NonStartableBrokerRule exporterBrokerRule = exporterIntegrationRule;
 
   @Before
   public void setUp() {
@@ -59,9 +62,38 @@ public abstract class AbstractElasticsearchExporterIntegrationTestCase {
       esClient = null;
     }
 
-    exporterBrokerRule.stop();
+    exporterIntegrationRule.stop();
     elastic.stop();
     configuration = null;
+  }
+
+  /**
+   * Starts the broker AND waits for all index templates to be present in Elastic before continuing.
+   */
+  protected void startBroker() {
+    startBrokerWithoutWaitingForIndexTemplates();
+    awaitIndexTemplatesCreation();
+  }
+
+  protected void startBrokerWithoutWaitingForIndexTemplates() {
+    exporterIntegrationRule.start();
+  }
+
+  protected void awaitIndexTemplatesCreation() {
+    final var expectedIndexTemplatesCount =
+        ValueTypeMapping.getAcceptedValueTypes().stream()
+            .filter(configuration::shouldIndexValueType)
+            .count();
+
+    // force exporting to ensure that we create the index templates, then wait for them to have been
+    // created
+    exporterBrokerRule.publishMessage("dummy", "");
+    Awaitility.await("until all indices have been created")
+        .atMost(Duration.ofMinutes(2))
+        .pollInterval(Duration.ofSeconds(1))
+        .until(
+            () -> esClient.getIndexTemplateCount(configuration.index.prefix),
+            size -> size >= expectedIndexTemplatesCount);
   }
 
   protected void assertIndexSettings() {
@@ -175,14 +207,14 @@ public abstract class AbstractElasticsearchExporterIntegrationTestCase {
     return configuration;
   }
 
-  protected static class ElasticsearchTestClient extends ElasticsearchClient {
+  public static class ElasticsearchTestClient extends ElasticsearchClient {
     private static final ObjectMapper MAPPER = new ObjectMapper();
 
     ElasticsearchTestClient(final ElasticsearchExporterConfiguration configuration) {
       super(configuration);
     }
 
-    GetSettingsForIndicesResponse getSettingsForIndices() {
+    public GetSettingsForIndicesResponse getSettingsForIndices() {
       final var request = new Request("GET", "/_all/_settings");
       try {
         final var response = client.performRequest(request);
@@ -199,7 +231,27 @@ public abstract class AbstractElasticsearchExporterIntegrationTestCase {
       }
     }
 
-    Map<String, Object> getDocument(final Record<?> record) {
+    public int getIndexTemplateCount(final String indexPrefix) {
+      final var request = new Request("GET", "/_index_template/" + indexPrefix + "*");
+      try {
+        final var response = client.performRequest(request);
+        final Map<String, List<Object>> indexTemplates =
+            MAPPER.readValue(response.getEntity().getContent(), new TypeReference<>() {});
+        return indexTemplates.getOrDefault("index_templates", new ArrayList<>()).size();
+      } catch (final ResponseException e) {
+        // you might get a 404 if there are no templates matching the pattern yet, but the response
+        // will be valid and of the same format
+        if (e.getResponse().getStatusLine().getStatusCode() == 404) {
+          return 0;
+        }
+
+        throw new ElasticsearchExporterException("Failed to get index template count", e);
+      } catch (final Exception e) {
+        throw new ElasticsearchExporterException("Failed to get index template count", e);
+      }
+    }
+
+    public Map<String, Object> getDocument(final Record<?> record) {
       final var request =
           new Request("GET", "/" + indexFor(record) + "/" + typeFor() + "/" + idFor(record));
       request.addParameter("routing", String.valueOf(record.getPartitionId()));

--- a/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterAuthenticationIT.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterAuthenticationIT.java
@@ -54,17 +54,14 @@ public class ElasticsearchExporterAuthenticationIT
     // given
     configuration = getDefaultConfiguration();
     exporterConfigurator.accept(configuration);
+    exporterBrokerRule.configure("es", ElasticsearchExporter.class, configuration);
+    esClient = createElasticsearchClient(configuration);
 
     // when
-    exporterBrokerRule.configure("es", ElasticsearchExporter.class, configuration);
-    exporterBrokerRule.start();
+    startBroker();
     exporterBrokerRule.performSampleWorkload();
 
     // then
-
-    // assert index settings for all created indices
-    esClient = createElasticsearchClient(configuration);
-
     // assert all records which where recorded during the tests where exported
     exporterBrokerRule.visitExportedRecords(
         r -> {

--- a/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterConfigurationIT.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterConfigurationIT.java
@@ -33,7 +33,7 @@ public class ElasticsearchExporterConfigurationIT
 
     // when
     exporterBrokerRule.configure("es", ElasticsearchExporter.class, configuration);
-    exporterBrokerRule.start();
+    startBroker();
     exporterBrokerRule.publishMessage("message", "123");
 
     // then
@@ -60,7 +60,7 @@ public class ElasticsearchExporterConfigurationIT
     exporterBrokerRule.configure("es", ElasticsearchExporter.class, configuration);
 
     // then
-    assertThatThrownBy(() -> exporterBrokerRule.start())
+    assertThatThrownBy(this::startBroker)
         .isInstanceOf(IllegalStateException.class)
         .getRootCause()
         .isInstanceOf(ExporterException.class)
@@ -83,7 +83,7 @@ public class ElasticsearchExporterConfigurationIT
     exporterBrokerRule.configure("es", ElasticsearchExporter.class, configuration);
 
     // then
-    assertThatThrownBy(() -> exporterBrokerRule.start())
+    assertThatThrownBy(this::startBroker)
         .isInstanceOf(IllegalStateException.class)
         .getRootCause()
         .isInstanceOf(ExporterException.class)

--- a/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterFaultToleranceIT.java
@@ -30,9 +30,10 @@ public class ElasticsearchExporterFaultToleranceIT
 
     // when
     exporterBrokerRule.configure("es", ElasticsearchExporter.class, configuration);
-    exporterBrokerRule.start();
+    startBrokerWithoutWaitingForIndexTemplates();
     exporterBrokerRule.publishMessage("message", "123");
     elastic.start();
+    awaitIndexTemplatesCreation();
 
     // then
     final var records =

--- a/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/records/ElasticsearchExporterDmnRecordIT.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/records/ElasticsearchExporterDmnRecordIT.java
@@ -34,7 +34,7 @@ public final class ElasticsearchExporterDmnRecordIT
     esClient = createElasticsearchClient(configuration);
 
     exporterBrokerRule.configure("es", ElasticsearchExporter.class, configuration);
-    exporterBrokerRule.start();
+    startBroker();
   }
 
   @Test

--- a/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/records/ElasticsearchExporterJobRecordIT.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/records/ElasticsearchExporterJobRecordIT.java
@@ -34,7 +34,7 @@ public class ElasticsearchExporterJobRecordIT
     esClient = createElasticsearchClient(configuration);
 
     exporterBrokerRule.configure("es", ElasticsearchExporter.class, configuration);
-    exporterBrokerRule.start();
+    startBroker();
   }
 
   @After

--- a/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/util/ElasticsearchContainer.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/util/ElasticsearchContainer.java
@@ -44,6 +44,9 @@ public class ElasticsearchContainer extends GenericContainer<ElasticsearchContai
 
   public ElasticsearchContainer(final String version) {
     super(DEFAULT_IMAGE + ":" + version);
+
+    // disable xpack by default to disable all these security warnings
+    withEnv("xpack.security.enabled", "false");
   }
 
   @Override

--- a/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/util/it/ExporterIntegrationRule.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/util/it/ExporterIntegrationRule.java
@@ -8,7 +8,6 @@
 package io.camunda.zeebe.exporter.util.it;
 
 import static io.camunda.zeebe.exporter.util.it.EmbeddedBrokerRule.TEST_RECORD_EXPORTER_ID;
-import static io.camunda.zeebe.test.util.record.RecordingExporter.processInstanceRecords;
 
 import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.ObjectMapper;
@@ -20,8 +19,6 @@ import io.camunda.zeebe.client.api.worker.JobWorker;
 import io.camunda.zeebe.exporter.api.Exporter;
 import io.camunda.zeebe.model.bpmn.BpmnModelInstance;
 import io.camunda.zeebe.protocol.record.Record;
-import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
-import io.camunda.zeebe.test.util.TestUtil;
 import io.camunda.zeebe.test.util.WorkloadGenerator;
 import io.camunda.zeebe.test.util.record.RecordingExporter;
 import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
@@ -111,7 +108,7 @@ import org.junit.runners.model.Statement;
  * NOTE: calls to the various configure methods are additive, so it is possible to configure more
  * than one exporter, as long as the IDs are different.
  */
-public class ExporterIntegrationRule extends ExternalResource {
+public class ExporterIntegrationRule extends ExternalResource implements NonStartableBrokerRule {
 
   private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
 
@@ -145,6 +142,7 @@ public class ExporterIntegrationRule extends ExternalResource {
    *
    * @return current broker configuration
    */
+  @Override
   public BrokerCfg getBrokerConfig() {
     return brokerRule.getBrokerCfg();
   }
@@ -152,6 +150,7 @@ public class ExporterIntegrationRule extends ExternalResource {
   /**
    * @return the currently configured exporters
    */
+  @Override
   public List<ExporterCfg> getConfiguredExporters() {
     return getBrokerConfig().getExporters().entrySet().stream()
         .filter(entry -> !entry.getKey().equals(TEST_RECORD_EXPORTER_ID))
@@ -162,6 +161,7 @@ public class ExporterIntegrationRule extends ExternalResource {
   /**
    * @return true if any exporter has been configured for the broker, false otherwise
    */
+  @Override
   public boolean hasConfiguredExporters() {
     return getConfiguredExporters().isEmpty();
   }
@@ -175,6 +175,7 @@ public class ExporterIntegrationRule extends ExternalResource {
    * @param <T> type of the configuration
    * @param <E> type of the exporter
    */
+  @Override
   public <T, E extends Exporter> ExporterIntegrationRule configure(
       final String id, final Class<E> exporterClass, final T configuration) {
     final Map<String, Object> arguments = convertConfigToMap(configuration);
@@ -189,6 +190,7 @@ public class ExporterIntegrationRule extends ExternalResource {
    * @param arguments the arguments to pass during configuration
    * @param <E> type of the exporter
    */
+  @Override
   public <E extends Exporter> ExporterIntegrationRule configure(
       final String id, final Class<E> exporterClass, final Map<String, Object> arguments) {
     final ExporterCfg config = new ExporterCfg();
@@ -196,6 +198,105 @@ public class ExporterIntegrationRule extends ExternalResource {
     config.setArgs(arguments);
 
     return configure(Collections.singletonMap(id, config));
+  }
+
+  /** Runs a sample workload on the broker, exporting several records of different types. */
+  @Override
+  public void performSampleWorkload() {
+    WorkloadGenerator.performSampleWorkload(clientRule.getClient());
+  }
+
+  /**
+   * Visits all exported records in the order they were exported.
+   *
+   * @param visitor record consumer
+   */
+  @Override
+  public void visitExportedRecords(final Consumer<Record<?>> visitor) {
+    RecordingExporter.getRecords().forEach(visitor);
+  }
+
+  /**
+   * Deploys the given process to the broker. Note that the filename must have the "bpmn" file
+   * extension, e.g. "resource.bpmn".
+   *
+   * @param process process to deploy
+   * @param filename resource name, e.g. "process.bpmn"
+   */
+  @Override
+  public void deployProcess(final BpmnModelInstance process, final String filename) {
+    clientRule
+        .getClient()
+        .newDeployResourceCommand()
+        .addProcessModel(process, filename)
+        .send()
+        .join();
+  }
+
+  /**
+   * Deploys the given classpath resource to the broker.
+   *
+   * @param classpathResource the resource to deploy
+   */
+  @Override
+  public void deployResourceFromClasspath(final String classpathResource) {
+    clientRule
+        .getClient()
+        .newDeployResourceCommand()
+        .addResourceFromClasspath(classpathResource)
+        .send()
+        .join();
+  }
+
+  /**
+   * Creates a process instance for the given process ID, with the given variables.
+   *
+   * @param processId BPMN process ID
+   * @param variables initial variables for the instance
+   * @return unique ID used to interact with the instance
+   */
+  @Override
+  public long createProcessInstance(final String processId, final Map<String, Object> variables) {
+    return clientRule
+        .getClient()
+        .newCreateInstanceCommand()
+        .bpmnProcessId(processId)
+        .latestVersion()
+        .variables(variables)
+        .send()
+        .join()
+        .getProcessInstanceKey();
+  }
+
+  /**
+   * Creates a new job worker that will handle jobs of type {@param type}.
+   *
+   * <p>Make sure to close the returned job worker.
+   *
+   * @param type type of the jobs to handle
+   * @param handler handler
+   * @return a new JobWorker
+   */
+  @Override
+  public JobWorker createJobWorker(final String type, final JobHandler handler) {
+    return clientRule.getClient().newWorker().jobType(type).handler(handler).open();
+  }
+
+  /**
+   * Publishes a new message to the broker.
+   *
+   * @param messageName name of the message
+   * @param correlationKey correlation key
+   */
+  @Override
+  public void publishMessage(final String messageName, final String correlationKey) {
+    clientRule
+        .getClient()
+        .newPublishMessageCommand()
+        .messageName(messageName)
+        .correlationKey(correlationKey)
+        .send()
+        .join();
   }
 
   /**
@@ -222,111 +323,6 @@ public class ExporterIntegrationRule extends ExternalResource {
     if (clientRule != null) {
       clientRule.destroyClient();
     }
-  }
-
-  /** Runs a sample workload on the broker, exporting several records of different types. */
-  public void performSampleWorkload() {
-    WorkloadGenerator.performSampleWorkload(clientRule.getClient());
-  }
-
-  /**
-   * Visits all exported records in the order they were exported.
-   *
-   * @param visitor record consumer
-   */
-  public void visitExportedRecords(final Consumer<Record<?>> visitor) {
-    RecordingExporter.getRecords().forEach(visitor);
-  }
-
-  /**
-   * Deploys the given process to the broker. Note that the filename must have the "bpmn" file
-   * extension, e.g. "resource.bpmn".
-   *
-   * @param process process to deploy
-   * @param filename resource name, e.g. "process.bpmn"
-   */
-  public void deployProcess(final BpmnModelInstance process, final String filename) {
-    clientRule
-        .getClient()
-        .newDeployResourceCommand()
-        .addProcessModel(process, filename)
-        .send()
-        .join();
-  }
-
-  /**
-   * Deploys the given classpath resource to the broker.
-   *
-   * @param classpathResource the resource to deploy
-   */
-  public void deployResourceFromClasspath(final String classpathResource) {
-    clientRule
-        .getClient()
-        .newDeployResourceCommand()
-        .addResourceFromClasspath(classpathResource)
-        .send()
-        .join();
-  }
-
-  /**
-   * Creates a process instance for the given process ID, with the given variables.
-   *
-   * @param processId BPMN process ID
-   * @param variables initial variables for the instance
-   * @return unique ID used to interact with the instance
-   */
-  public long createProcessInstance(final String processId, final Map<String, Object> variables) {
-    return clientRule
-        .getClient()
-        .newCreateInstanceCommand()
-        .bpmnProcessId(processId)
-        .latestVersion()
-        .variables(variables)
-        .send()
-        .join()
-        .getProcessInstanceKey();
-  }
-
-  /**
-   * Creates a new job worker that will handle jobs of type {@param type}.
-   *
-   * <p>Make sure to close the returned job worker.
-   *
-   * @param type type of the jobs to handle
-   * @param handler handler
-   * @return a new JobWorker
-   */
-  public JobWorker createJobWorker(final String type, final JobHandler handler) {
-    return clientRule.getClient().newWorker().jobType(type).handler(handler).open();
-  }
-
-  /**
-   * Publishes a new message to the broker.
-   *
-   * @param messageName name of the message
-   * @param correlationKey correlation key
-   */
-  public void publishMessage(final String messageName, final String correlationKey) {
-    clientRule
-        .getClient()
-        .newPublishMessageCommand()
-        .messageName(messageName)
-        .correlationKey(correlationKey)
-        .send()
-        .join();
-  }
-
-  /**
-   * Blocks and wait until the process identified by the key has been completed.
-   *
-   * @param processInstanceKey ID of the process
-   */
-  public void awaitProcessCompletion(final long processInstanceKey) {
-    TestUtil.waitUntil(
-        () ->
-            processInstanceRecords(ProcessInstanceIntent.ELEMENT_COMPLETED)
-                .filter(r -> r.getKey() == processInstanceKey)
-                .exists());
   }
 
   private Properties newClientProperties() {

--- a/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/util/it/NonStartableBrokerRule.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/util/it/NonStartableBrokerRule.java
@@ -1,0 +1,56 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.exporter.util.it;
+
+import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
+import io.camunda.zeebe.broker.system.configuration.ExporterCfg;
+import io.camunda.zeebe.client.api.worker.JobHandler;
+import io.camunda.zeebe.client.api.worker.JobWorker;
+import io.camunda.zeebe.exporter.AbstractElasticsearchExporterIntegrationTestCase;
+import io.camunda.zeebe.exporter.api.Exporter;
+import io.camunda.zeebe.model.bpmn.BpmnModelInstance;
+import io.camunda.zeebe.protocol.record.Record;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * This interface is a temporary workaround to prevent actual tests from starting a broker manually.
+ * Instead, you should use {@link AbstractElasticsearchExporterIntegrationTestCase#startBroker()}.
+ *
+ * <p>As all of these utility classes will be removed by the end of Q2-2022, it's an OK trade-off to
+ * avoid rewriting most of the tests (which will be rewritten differently anyway).
+ */
+public interface NonStartableBrokerRule {
+
+  BrokerCfg getBrokerConfig();
+
+  List<ExporterCfg> getConfiguredExporters();
+
+  boolean hasConfiguredExporters();
+
+  <T, E extends Exporter> ExporterIntegrationRule configure(
+      String id, Class<E> exporterClass, T configuration);
+
+  <E extends Exporter> ExporterIntegrationRule configure(
+      String id, Class<E> exporterClass, Map<String, Object> arguments);
+
+  void performSampleWorkload();
+
+  void visitExportedRecords(Consumer<Record<?>> visitor);
+
+  void deployProcess(BpmnModelInstance process, String filename);
+
+  void deployResourceFromClasspath(String classpathResource);
+
+  long createProcessInstance(String processId, Map<String, Object> variables);
+
+  JobWorker createJobWorker(String type, JobHandler handler);
+
+  void publishMessage(String messageName, String correlationKey);
+}

--- a/test-util/src/main/java/io/camunda/zeebe/test/util/record/RecordingExporter.java
+++ b/test-util/src/main/java/io/camunda/zeebe/test/util/record/RecordingExporter.java
@@ -97,6 +97,7 @@ public final class RecordingExporter implements Exporter {
   public static void reset() {
     LOCK.lock();
     try {
+      maximumWaitTime = DEFAULT_MAX_WAIT_TIME;
       RECORDS.clear();
     } finally {
       LOCK.unlock();

